api/krusty/component_test.go

```diff
@@ -0,0 +1,441 @@
+- deploy.yaml
+configMapGenerator:
+- name: my-configmap
+  literals:	
+    - testValue=1
+    - otherValue=10
+`)
+	th.WriteF("/app/base/deploy.yaml", `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: storefront
+spec:
+  replicas: 1
+`)
+}
+
+func writeComponentPatch(th kusttest_test.Harness) {
+	th.WriteF("/app/patch/kustomization.yaml", `
+apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+namePrefix: patched-
+replicas:
+- name: storefront
+  count: 3
+resources:
+  - stub.yaml
+configMapGenerator:
+- name: my-configmap
+  behavior: merge
+  literals:	
+    - testValue=2
+    - patchValue=5
+`)
+	th.WriteF("/app/patch/stub.yaml", `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: stub
+spec:
+  replicas: 1
+`)
+}
+
+func writeComponentProd(th kusttest_test.Harness) {
+	th.WriteK("/app/prod", `
+resources:
+- ../base
+- db
+
+components:
+- ../patch
+`)
+	th.WriteF("/app/prod/db", `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: db
+spec:
+  type: Logical
+`)
+}
+
+// Components are inserted into the resource hierarchy as the parent of those
+// resources that come before it in the resources list of the parent Kustomization.
+func TestBasicComponent(t *testing.T) {
+	th := kusttest_test.MakeHarness(t)
+	writeComponentBase(th)
+	writeComponentPatch(th)
+	writeComponentProd(th)
+	m := th.Run("/app/prod", th.MakeDefaultOptions())
+	th.AssertActualEqualsExpected(m, `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-storefront
+spec:
+  replicas: 3
+---
+apiVersion: v1
+data:
+  otherValue: "10"
+  patchValue: "5"
+  testValue: "2"
+kind: ConfigMap
+metadata:
+  annotations: {}
+  labels: {}
+  name: patched-my-configmap-5g55h28cdh
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-db
+spec:
+  type: Logical
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-stub
+spec:
+  replicas: 1
+`)
+}
+
+func TestMultipleComponents(t *testing.T) {
+	th := kusttest_test.MakeHarness(t)
+	writeComponentBase(th)
+	writeComponentPatch(th)
+	writeComponentProd(th)
+	th.WriteF("/app/additionalpatch/kustomization.yaml", `
+apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+configMapGenerator:
+- name: my-configmap
+  behavior: merge
+  literals:	
+    - otherValue=9
+`)
+	th.WriteK("/app/prod", `
+resources:
+- ../base
+- db
+
+components:
+- ../patch
+- ../additionalpatch
+`)
+	m := th.Run("/app/prod", th.MakeDefaultOptions())
+	th.AssertActualEqualsExpected(m, `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-storefront
+spec:
+  replicas: 3
+---
+apiVersion: v1
+data:
+  otherValue: "9"
+  patchValue: "5"
+  testValue: "2"
+kind: ConfigMap
+metadata:
+  annotations: {}
+  labels: {}
+  name: patched-my-configmap-9fddc87cdk
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-db
+spec:
+  type: Logical
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: patched-stub
+spec:
+  replicas: 1
+`)
+}
+
+func TestNestedComponents(t *testing.T) {
```

Comment:
nice

Reply:
:) I like how the (recursive) implementation of Kustomize made this just work!